File: designer/testSrc/com/android/tools/idea/uibuilder/property/editors/support/TextAppearanceEnumSupportTest.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.uibuilder.property.editors.support;

import static com.android.SdkConstants.ANDROID_URI;
import static com.android.SdkConstants.APPCOMPAT_LIB_ARTIFACT_ID;
import static com.android.SdkConstants.ATTR_STYLE;
import static com.android.SdkConstants.CHECK_BOX;
import static com.android.SdkConstants.CONSTRAINT_LAYOUT;
import static com.android.SdkConstants.TEXT_VIEW;
import static com.android.tools.idea.uibuilder.property.editors.support.TextAppearanceEnumSupport.TEXT_APPEARANCE_PATTERN;
import static com.google.common.truth.Truth.assertThat;

import com.android.tools.idea.common.SyncNlModel;
import com.android.tools.idea.common.fixtures.ComponentDescriptor;
import com.android.tools.idea.common.fixtures.ModelBuilder;
import com.android.tools.idea.common.model.NlComponent;
import com.android.tools.idea.testing.AndroidProjectRule;
import com.android.tools.idea.testing.Dependencies;
import com.android.tools.idea.uibuilder.NlModelBuilderUtil;
import com.android.tools.idea.uibuilder.property.NlPropertyItem;
import com.google.common.collect.ImmutableList;
import com.intellij.testFramework.EdtRule;
import com.intellij.testFramework.RunsInEdt;
import com.intellij.util.xml.XmlName;
import java.util.List;
import java.util.regex.Matcher;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class TextAppearanceEnumSupportTest {

  @Rule
  public final AndroidProjectRule myProjectRule = AndroidProjectRule.withSdk().initAndroid(true);

  @Rule
  public final EdtRule myEdtRule = new EdtRule();

  private SyncNlModel myModel;
  private NlComponent myLayout;

  @Before
  public void setUp() {
    addProjectStyles();
    myModel = createModel().build();
    myLayout = myModel.getComponents().get(0);
  }

  @After
  public void tearDown() {
    myModel = null;
    myLayout = null;
  }

  @Test
  public void testTextAppearancePattern() {
    checkTextAppearancePattern("TextAppearance", true, null);
    checkTextAppearancePattern("TextAppearance.Small", true, "Small");
    checkTextAppearancePattern("@android:style/TextAppearance.Material.Small", true, "Material.Small");
    checkTextAppearancePattern("@style/TextAppearance.AppCompat.Small", true, "AppCompat.Small");
    checkTextAppearancePattern("WhatEver", false, null);
  }

  private static void checkTextAppearancePattern(@NotNull String value, boolean expectedMatch, @Nullable String expectedMatchValue) {
    Matcher matcher = TEXT_APPEARANCE_PATTERN.matcher(value);
    assertThat(matcher.matches()).isEqualTo(expectedMatch);
    if (expectedMatch) {
      assertThat(matcher.group(5)).isEqualTo(expectedMatchValue);
    }
  }

  @RunsInEdt
  @Test
  public void testFindPossibleValues() {
    Dependencies.INSTANCE.add(myProjectRule.fixture, APPCOMPAT_LIB_ARTIFACT_ID);
    NlComponent textView = myLayout.getChild(0);
    NlPropertyItem property = NlPropertyItem.create(new XmlName(ATTR_STYLE, ANDROID_URI), null, ImmutableList.of(textView), null);
    TextAppearanceEnumSupport support = new TextAppearanceEnumSupport(property);
    List<ValueWithDisplayString> values = support.getAllValues();
    assertThat(values.subList(0, 7)).containsExactly(
      new ValueWithDisplayString("MyOwnStyle", "@style/MyOwnStyle"),
      new ValueWithDisplayString("MyOwnStyle.Medium", "@style/MyOwnStyle.Medium"),
      new ValueWithDisplayString("TextAppearance", "@style/TextAppearance"),
      ValueWithDisplayString.SEPARATOR,
      new ValueWithDisplayString("AppCompat", "@style/TextAppearance.AppCompat"),
      new ValueWithDisplayString("AppCompat.Body1", "@style/TextAppearance.AppCompat.Body1"),
      new ValueWithDisplayString("AppCompat.Body2", "@style/TextAppearance.AppCompat.Body2")).inOrder();

    assertThat(support.getAllValues()).containsAllOf(
      new ValueWithDisplayString("TextAppearance", "@android:style/TextAppearance"),
      new ValueWithDisplayString("Material", "@android:style/TextAppearance.Material"),
      new ValueWithDisplayString("Material.Small", "@android:style/TextAppearance.Material.Small"),
      new ValueWithDisplayString("AppCompat", "@style/TextAppearance.AppCompat"),
      new ValueWithDisplayString("AppCompat.Display1", "@style/TextAppearance.AppCompat.Display1"),
      new ValueWithDisplayString("AppCompat.Display2", "@style/TextAppearance.AppCompat.Display2"),
      new ValueWithDisplayString("AppCompat.Display3", "@style/TextAppearance.AppCompat.Display3"),
      new ValueWithDisplayString("AppCompat.Display4", "@style/TextAppearance.AppCompat.Display4"),
      new ValueWithDisplayString("Widget.TextView", "@android:style/TextAppearance.Widget.TextView"));
  }

  private void addProjectStyles() {
    String styles =
      "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n" +
      "<resources xmlns:android=\"http://schemas.android.com/apk/res/android\">\n" +
      "    <style name=\"MyOwnStyle\" parent=\"android:TextAppearance\"/>\n" +
      "    <style name=\"MyOwnStyle.Medium\"/>\n" +
      "    <style name=\"TextAppearance\" parent=\"android:TextAppearance\"/>\n" +
      "</resources>\n";
    myProjectRule.fixture.addFileToProject("res/values/styles.xml", styles);
  }

  private ModelBuilder createModel() {
    return NlModelBuilderUtil.model(
      myProjectRule,
      "layout",
      "constraint.xml",
      new ComponentDescriptor(CONSTRAINT_LAYOUT.defaultName())
        .id("@id/root")
        .withBounds(0, 0, 2000, 2000)
        .width("1000dp")
        .height("1000dp")
        .withAttribute("android:padding", "20dp")
        .children(
          new ComponentDescriptor(CHECK_BOX)
            .id("@id/button")
            .withBounds(100, 100, 200, 40)
            .width("100dp")
            .height("20dp"),
          new ComponentDescriptor(TEXT_VIEW)
            .id("@id/button")
            .withBounds(200, 100, 200, 40)
            .width("100dp")
            .height("20dp")
        ));
  }
}

File: designer/testSrc/com/android/tools/idea/uibuilder/property2/support/TextAppearanceEnumSupportTest.kt
/*
 * Copyright (C) 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.tools.idea.uibuilder.property2.support

import com.android.SdkConstants.*
import com.android.tools.idea.uibuilder.property2.NelePropertyType
import com.android.tools.idea.uibuilder.property2.testutils.EnumValueUtil
import com.android.tools.idea.uibuilder.property2.testutils.SupportTestUtil
import com.google.common.truth.Truth
import org.jetbrains.android.AndroidTestCase

private const val APPCOMPAT_TEXT_APPEARANCES = """
<resources>
    <style name="Base.TextAppearance.AppCompat" parent="android:TextAppearance"/>
    <style name="TextAppearance.AppCompat" parent="Base.TextAppearance.AppCompat"/>
    <style name="TextAppearance.AppCompat.Small"/>
    <style name="TextAppearance.AppCompat.Medium"/>
    <style name="TextAppearance.AppCompat.Large"/>
    <style name="TextAppearance.AppCompat.Body1"/>
    <style name="TextAppearance.AppCompat.Body2"/>
    <style name="TextAppearance.AppCompat.Display1"/>
    <style name="TextAppearance.AppCompat.Display2"/>
    <style name="TextAppearance.AppCompat.Display3"/>
    <style name="TextAppearance.AppCompat.Display4"/>
</resources>"""

private const val PROJECT_TEXT_APPEARANCES = """
<resources>
    <style name="MyTextStyle" parent="android:TextAppearance"/>
    <style name="TextAppearance.Blue" parent="MyTextStyle"/>
</resources>"""

class TextAppearanceEnumSupportTest: AndroidTestCase() {

  override fun setUp() {
    super.setUp()
    myFixture.addFileToProject("res/values/project_styles.xml", PROJECT_TEXT_APPEARANCES)
  }

  fun testTextViewTextAppearanceWithAppCompat() {
    // setup
    myFixture.addFileToProject("res/values/appcompat_styles.xml", APPCOMPAT_TEXT_APPEARANCES)
    val util = SupportTestUtil(testRootDisposable, myFacet, myFixture, TEXT_VIEW)
    val property = util.makeProperty(ANDROID_URI, ATTR_TEXT_APPEARANCE, NelePropertyType.STYLE)
    EnumValueUtil.patchLibraryNameOfAllAppCompatStyles(property)

    // test
    val support = TextAppearanceEnumSupport(property)
    val values = support.values
    val expectedProjectValues = listOf("@style/TextAppearance.Blue", "@style/MyTextStyle")
    val expectedProjectDisplayValues = listOf("Blue", "MyTextStyle")
    val expectedAppCompatValues = listOf(
        "@style/TextAppearance.AppCompat.Body1",
        "@style/TextAppearance.AppCompat.Body2",
        "@style/TextAppearance.AppCompat.Display1",
        "@style/TextAppearance.AppCompat.Display2",
        "@style/TextAppearance.AppCompat.Display3",
        "@style/TextAppearance.AppCompat.Display4",
        "@style/TextAppearance.AppCompat.Large",
        "@style/TextAppearance.AppCompat.Medium",
        "@style/TextAppearance.AppCompat.Small")
    val expectedAppCompatDisplayValues = listOf(
        "Body1",
        "Body2",
        "Display1",
        "Display2",
        "Display3",
        "Display4",
        "Large",
        "Medium",
        "Small")
    var index = 0
    index = EnumValueUtil.checkSection(values, index, PROJECT_HEADER, 2, expectedProjectValues, expectedProjectDisplayValues)
    index = EnumValueUtil.checkSection(values, index, APPCOMPAT_HEADER, 9, expectedAppCompatValues, expectedAppCompatDisplayValues)
    Truth.assertThat(index).isEqualTo(-1)
  }

  fun testTextViewTextAppearanceWithoutAppCompat() {
    // setup
    val util = SupportTestUtil(testRootDisposable, myFacet, myFixture, TEXT_VIEW)
    val property = util.makeProperty(ANDROID_URI, ATTR_TEXT_APPEARANCE, NelePropertyType.STYLE)

    // test
    val support = TextAppearanceEnumSupport(property)
    val values = support.values
    val expectedProjectValues = listOf("@style/TextAppearance.Blue", "@style/MyTextStyle")
    val expectedProjectDisplayValues = listOf("Blue", "MyTextStyle")
    val expectedAndroidValues = listOf(
        "@android:style/TextAppearance.Material.Body1",
        "@android:style/TextAppearance.Material.Body2",
        "@android:style/TextAppearance.Material.Display1",
        "@android:style/TextAppearance.Material.Display2",
        "@android:style/TextAppearance.Material.Display3",
        "@android:style/TextAppearance.Material.Display4",
        "@android:style/TextAppearance.Material.Large",
        "@android:style/TextAppearance.Material.Medium",
        "@android:style/TextAppearance.Material.Small")
    val expectedAndroidDisplayValues = listOf(
        "Body1",
        "Body2",
        "Display1",
        "Display2",
        "Display3",
        "Display4",
        "Large",
        "Medium",
        "Small")
    var index = 0
    index = EnumValueUtil.checkSection(values, index, PROJECT_HEADER, 2, expectedProjectValues, expectedProjectDisplayValues)
    index = EnumValueUtil.checkSection(values, index, ANDROID_HEADER, 9, expectedAndroidValues, expectedAndroidDisplayValues)
    Truth.assertThat(index).isEqualTo(-1)
  }
}
